src/plugin-client.ts

```diff
@@ -11,26 +11,47 @@
 
 'use strict';
 
-const AddonManagerProxy = require('./addon-manager-proxy');
-const {MessageType} = require('./constants');
-const Deferred = require('./deferred');
-const EventEmitter = require('events');
-const IpcSocket = require('./ipc');
-
-class PluginClient extends EventEmitter {
-  constructor(pluginId, {verbose} = {}) {
+import { AddonManagerProxy } from './addon-manager-proxy';
+import { MessageType } from './constants';
+import { Deferred } from './deferred';
+import { EventEmitter } from 'events';
+import { IpcSocket } from './ipc';
+import WebSocket from 'ws';
+
+export class PluginClient extends EventEmitter {
+  private verbose: boolean;
+  private deferredReply?: Deferred<AddonManagerProxy, void>;
+  private logPrefix: string;
+  private gatewayVersion?: string;
+  private userProfile: any;
+  private preferences: any;
+  private addonManager?: AddonManagerProxy;
+  private ipcSocket?: IpcSocket;
+  private ws?: WebSocket;
+
+  constructor(private pluginId: string, { verbose }: any = {}) {
     super();
-    this.pluginId = pluginId;
-    this.verbose = verbose;
-    this.deferredReply = null;
+    this.verbose = !!verbose;
     this.logPrefix = `PluginClient(${this.pluginId}):`;
   }
 
-  onMsg(msg) {
+  getGatewayVersion() {
+    return this.gatewayVersion;
+  }
+
+  getUserProfile() {
+    return this.userProfile;
+  }
+
+  getPreferences() {
+    return this.preferences;
+  }
+
+  onMsg(msg: any) {
     this.verbose &&
       console.log(this.logPrefix, 'rcvd ManagerMsg:', msg);
 
-    if (msg.messageType === MessageType.PLUGIN_REGISTER_RESPONSE) {
+    if (msg.messageType === (<any>MessageType).PLUGIN_REGISTER_RESPONSE) {
```

Comment:
Why does `MessageType` have to be cast to `any`?

Reply:
Because it does not have any fields at compile time.

Reply:
Maybe we could generate it at compile time.

Reply:
I am for "the less any, the better; no any is the best" way :-) But I don't thinks that should be in the scope of this PR as it is already massive.